### PR TITLE
🚨🚨🚨 [`SPM`] Finish fix spm models 🚨🚨🚨

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Modifies `Llama` and `T5` other sentencepiece based tokenizer will follow. 

Previous behaviour is always possible with ` tokenizer = AutoTokenizer.from_pretrained("huggyllama/llama-7b", legacy = True)`
## The goal of `transformers`'s wrapping around `sentencepiece`
To clarify, we want to:
- be able to choose the behaviour of the special/added tokens. This means handling the `stripping`, encoding and decoding of such tokens
- allow users to easily add new tokens, with `tokenenizer.add_tokens(...)` instead of having to load the protobuf file, modify the vocab, save it and reload the sentencepiece processor. 

## The current and past problems with our wrappers
Let's use both T5 and Llama as reference models. Currently, we do not mimic the behaviour of adding words to the actual `sentencepiece` vocabulary. This is an issue for anyone expecting (and rightfully) that adding tokens does not modify the behaviour of the model. 

### Adding a word to sentencepiece's vocab
This can be done using: ([source](https://github.com/google/sentencepiece/issues/121#issuecomment-400362011))

```python 
>>> # wget https://huggingface.co/huggyllama/llama-7b/resolve/main/tokenizer.model
>>> from sentencepiece import sentencepiece_model_pb2 as model
>>> import sentencepiece as spm
>>> sp_model = model.ModelProto()
>>> sp_model.ParseFromString(open('tokenizer.model', 'rb').read())
>>> token = "your_token"
>>> sp_model.pieces.add(piece=f"{token}",score=0.0,type=model.ModelProto.SentencePiece.USER_DEFINED,)
>>> with open('new.model', 'wb') as f:
...     f.write(sp_model.SerializeToString())
```
then load the `sp_model`:
```python 
>>> sp_model = spm.SentencePieceProcessor()
>>> sp_model.Load('new.model')
```

Then, try the following :
```python 
>>> sp_model.encode("your_tokenHello", out_type=str)
["_", "your_token", "Hello"]
```

### Adding a word to a `PretrainedTokenizer
This can be done using `tokenizer.add_tokens(["your_token"])`. It is a lot simpler indeed. 
But the output you will get is: 
```python 
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("huggyllama/llama-7b", legacy = True, use_fast = False)
>>> tokenizer.add_tokens(["your_token"])
>>> tokenizer.tokenize("your_tokenHello")
["your_token", "_Hello"]

>>> tokenizer = AutoTokenizer.from_pretrained("huggyllama/llama-7b", legacy = False, use_fast = False)
>>> tokenizer.add_tokens(["your_token"])
>>> tokenizer.tokenize("your_tokenHello")
["your_token", "Hello"]
```

This is because we always split the text on the added tokens, and give the text on the left and right to the `sentencepiece` model. But, most sentencepiece models add a prefix space `_` (the `SPIECE_UNDERLINE` character). Thus, when the `transformers` tokenizers splits `"your_tokenHello"`, it encode `your_token` with the `tokenizer.added_tokens_encoder` and thus does not add a prefix space, and then encode `Hello` with the sentencepiece model, which adds a prefix space and thus outputs `_Hello`. 

Other missmatches: 

```python
# t5-base tokenizer
>>> tokenizer.encode("<extra_id_0>. Hello", add_special_tokens = False)
[32099, 3, 5, 8774] # ['<extra_id_0>', ' ▁', '.', '▁Hello']
# seqio.SentencePieceVocabulary(vocab_path, extra_ids = 300)
>>> processor.encode("<extra_id_0>. Hello")
[32099, 5, 8774] # ['<extra_id_0>', '.', '▁Hello']

``` 

TLDR; this shows the only way we can actually and properly handle added tokens and sentencepiece. We have to disable automatic prefix addition, and always encode with a token that is part of the vocab at the beginning to properly encode the first token, whether it has a prefix space or not. Yes this is dirty and sad, but the previous fix was removing the extra space, which was cleaner but had a corner cases #25176. 

### The same issue happens with fast tokenizers:
```python 
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("huggyllama/llama-7b", use_fast = True)
>>> tokenizer.add_tokens(["your_token"])
>>> tokenizer.tokenize("your_tokenHello")
["_your_token", "Hello"]

>>> tokenizer.add_tokens(["your_token_special"], True)
>>> tokenizer.tokenize("your_token_specialHello")
['your_token_special', '▁Hello']
```

### Another issue 😈 

So, here, the issue is that before the special token, even if there is no `rstrip` or `lstrip` (both are set to False), we have very strange behaviours:

```python 
>>> tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", use_fast = True)

>>> tokenizer.tokenize("<s>inform<s>")
# prefix space is eaten
['<s>', '▁inform', '<s>']

>>> tokenizer.tokenize("<s>inform <s>")
# prefix space is not eaten for the second <s>
['<s>', '▁inform', '▁', '<s>']

>>> tokenizer.tokenize(" <s>inform <s>")
# prefix space is not eaten for the second <s>
['▁▁', '<s>', '▁inform', '▁', '<s>']

>>> tokenizer.tokenize(" <s>inform <s> ")
# prefix space is not eaten for the first <s>, extra space added (known)
['▁▁', '<s>', '▁inform', '▁', '<s>', '▁▁']

>>> tokenizer.tokenize("inform <s> ")
# prefix space is added to inform
['▁inform', '▁', '<s>', '▁▁']
```
Note that `tokenizer.convert_tokens_to_ids("▁▁") = 259` while `tokenizer.convert_tokens_to_ids("▁") = 29871`
Also if we add a prefix space to special tokens the beginning, we are probably gonna break a lot of things

### CODE REVIEW

File: tests/models/t5/test_tokenization_t5.py
Comment (ArthurZucker) [inline tests/models/t5/test_tokenization_t5.py:1]: The previous test values were not really good, with this update it makes more sense

File: tests/models/llama/test_tokenization_llama.py [hunk 1]
```diff
@@ -534,15 +555,19 @@ def test_remove_extra_whitespaces(self):
         input_ids = self.tokenizer.encode("       . Hello")
         self.assertEqual(input_ids, [7, 4, 156, 86, 20])
         sp_encode = self.tokenizer.sp_model.encode("       . Hello")
-        self.assertEqual(input_ids, sp_encode)
+        self.assertEqual(input_ids, [7] + sp_encode)
```
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:558]: Manually add the `_` (spiece underline)

File: src/transformers/models/t5/tokenization_t5.py [hunk 2]
```diff
@@ -349,15 +366,10 @@ def _tokenize(self, text, **kwargs):
...
-                text = text[1:]
-
-        tokens = self.sp_model.encode(text, out_type=str)
-
-        if not self.legacy and not is_first and not text.startswith(" ") and tokens[0].startswith(SPIECE_UNDERLINE):
-            tokens = ([tokens[0][1:]] if len(tokens[0]) > 1 else []) + tokens[1:]
-        return tokens
+            text = self.unk_token + text
+            tokens = self.sp_model.encode(text, out_type=str)
+            return tokens[self.unk_token_length :]
```
Comment (ArthurZucker) [inline src/transformers/models/t5/tokenization_t5.py:371]: That's the Hack: 
- all spm models have a tokenizer. Whether or not it is in the sentencepiece vocab does not matter. 
- we need to do this because since `add_dummy_prefix = False` the sentencpiece model always ALWAYS strips any `SPIECE_UNDERLINE`. So `sp_model.encode(SPIECE_UNDERLINE + "Hello" , out_type=str)` will give [`Hel`,`llo`]  instead of [`_Hel`, `llo`]. 
- previously, we removed added extra space. This is okay, but fails for words that should be split like `inform`. What happened before was that we would tokenize as `_inform` then remove `_` and we have `inform`. But, the actual tokenization of `inform` is `in`,`form` and `inform` is not part of the vocab!

**Review by ArthurZucker — COMMENTED** [review]
Help for reviewers

**Review by sgugger — APPROVED** [review]
This makes sense to me. As usual with breaking changes, can you put three 🚨 in the title and show in the description how to enable the past behavior for users who want it?

File: src/transformers/models/llama/tokenization_llama.py [hunk 3]
```diff
@@ -24,7 +24,9 @@
 from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple
 
 import sentencepiece as spm
+from sentencepiece import SentencePieceProcessor
```
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:27]: Out of interest - why import directly here instead of using `spm.SentencePieceProcessor` as before?

File: src/transformers/models/llama/tokenization_llama.py [hunk 4]
```diff
@@ -187,15 +209,10 @@ def _tokenize(self, text):
...
-
-        tokens = self.sp_model.encode(text, out_type=str)
-
-        if not self.legacy and not is_first and not text.startswith(" ") and tokens[0].startswith(SPIECE_UNDERLINE):
-            tokens = ([tokens[0][1:]] if len(tokens[0]) > 1 else []) + tokens[1:]
-        return tokens
+            text = self.unk_token + text
+            tokens = self.sp_model.encode(text, out_type=str)
+            return tokens[self.unk_token_length :]
+        return self.sp_model.encode(text, out_type=str)
```
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:215]: nit: as the non-legacy code is the 'main' code now, it would be cleaner to have an early return in the case of `self.legacy=True` and the main logic unindented

```suggestion
        if self.legacy:
            return self.sp_model.encode(text, out_type=str) 
        text = self.unk_token + text
        tokens = self.sp_model.encode(text, out_type=str)
        return tokens[self.unk_token_length :]
```

File: src/transformers/models/llama/tokenization_llama.py [hunk 5]
```diff
@@ -209,6 +226,8 @@ def _convert_id_to_token(self, index):
     def convert_tokens_to_string(self, tokens):
         """Converts a sequence of tokens (string) in a single string."""
         current_sub_tokens = []
+        # since we manually add the prefix space, we have to remove it
+        tokens[0] = tokens[0].strip(SPIECE_UNDERLINE)
```
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:230]: It would be a bit safer using `lstrip` here - as I believe it's always a prefix we're removing 
```suggestion
        tokens[0] = tokens[0].lstrip(SPIECE_UNDERLINE)
```

File: tests/models/t5/test_tokenization_t5.py [hunk 6]
```diff
@@ -410,9 +410,11 @@ class CommonSpmIntegrationTests(unittest.TestCase):
 
     @classmethod
     def setUpClass(cls):
-        tokenizer = T5Tokenizer(SAMPLE_VOCAB, extra_ids=0, legacy=False)
-        tokenizer.add_special_tokens({"additional_special_tokens": ["<extra_id_0>"]})
+        tokenizer = T5Tokenizer(SAMPLE_VOCAB, extra_ids=1, legacy=False)
+        # tokenizer.add_tokens("<extra_id_0>", special_tokens = True)
+        # tokenizer._additional_special_tokens = ["<extra_id_0>"]
```
Comment (amyeroberts) [inline tests/models/t5/test_tokenization_t5.py:415]: To remove?

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for working on this v. tricky problem 🤗

File: src/transformers/models/llama/tokenization_llama.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py:27]: No particular reason, will revert 😉

File: src/transformers/models/llama/tokenization_llama.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py:231]: Yes! Though there will never be a space on the left, still more readable thanks!

**Comment by ArthurZucker** [comment]
Will fix the prefixing of special tokens!

File: tests/models/llama/test_tokenization_llama.py [hunk 7]
```diff
@@ -297,6 +297,10 @@ def test_picklable(self):
     def test_pickle_subword_regularization_tokenizer(self):
         pass
 
+    @unittest.skip("worker 'gw4' crashed on CI, passing locally.")
```
Comment (sgugger) [inline tests/models/llama/test_tokenization_llama.py:300]: This means the CI gets out of RAM FYI

**Comment by faaany** [comment]
@ArthurZucker any update to this PR?

**Comment by ArthurZucker** [comment]
Hey @faaany, I am updating it right now!

**Comment by ArthurZucker** [comment]
Reverted the changes as adding proper support for `add_prefix_space` is actully questionable. The usecase is already wrong as you should be reverse looking for ids not strings. See #24846 (adding prefix space was almost never done properly as the decoders were not updated as well)

**Comment by ArthurZucker** [comment]
pinging @sgugger for a final review !

**Review by sgugger — APPROVED** [review]
LGTM! Would be nice to make the two tests skipped smaller so that they pass.

**Comment by ArthurZucker** [comment]
Will do so in a follow up PR!

**Comment by zhacmsra** [comment]
I have added the legacy=True as "enc = AutoTokenizer.from_pretrained(model_path, legacy = True, use_fast=False)" but I have still gotten an error, which is a "Not a string" error. Anyone can give a hint what is going on here?

截屏2023-08-20 21 40 35

**Comment by ArthurZucker** [comment]
@zhacmsra the issue is in loading the vocabulary file, not 100% sure it's related to this. Can you open a new Issue with a reproducer please?

**Comment by zhacmsra** [comment]
> 

Hi Arthur, you are correct. I figured out that it is not related to this PR. The networking problem broke the input model and result in error inputs. Sorry for this trouble. Thank you for the kind and timely response.

File: src/transformers/models/llama/tokenization_llama.py [hunk 8]
```diff
@@ -129,22 +131,42 @@ def __init__(
...
             legacy=legacy,
             **kwargs,
         )
         if legacy is None:
             logger.warning_once(
-                f"You are using the default legacy behaviour of the {self.__class__}. This means that tokens that come after special tokens will not be properly handled. We recommend you to"
-                " read the related pull request available at https://github.com/huggingface/transformers/pull/24565, and set the legacy attribute accordingly."
+                f"You are using the default legacy behaviour of the {self.__class__}. If you see this, DO NOT PANIC! This is"
+                " expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you."
+                " If you want to use the new behaviour, set `legacy=True`. This should only be set if you understand what it"
```
Comment (CoaLee) [inline src/transformers/models/llama/tokenization_llama.py:142]: Wrong log message. This part should be: 
```
..., set `legacy=False`. ...
```

Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py:142]: this was updated on main

Comment (CoaLee) [reply src/transformers/models/llama/tokenization_llama.py:142]: 😯 Should have checked that. Thanks.
